src/full/Agda/Compiler/Common.hs

```diff
@@ -51,13 +51,19 @@ instance Monoid IsMain where
 
 doCompile :: Monoid r => (IsMain -> Interface -> TCM r) -> IsMain -> Interface -> TCM r
 doCompile f isMain i = do
-  -- The Agda.Primitive module is implicitly assumed to be always imported,
-  -- even though it not necesseraly occurs in iImportedModules.
-  -- TODO: there should be a better way to get hold of Agda.Primitive?
-  [agdaPrimInter] <- filter (("Agda.Primitive"==) . prettyShow . iModuleName)
-    . map miInterface . Map.elems
-      <$> getVisitedModules
-  flip evalStateT Set.empty $ mappend <$> doCompile' f NotMain agdaPrimInter <*> doCompile' f isMain i
+  -- The Agda.Primitive module is only loaded if the
+  -- --no-load-primitives flag was not given, i.e. if optLoadPrimitives
+  -- is True. If --no-load-primitives was given, then we won't find an
+  -- interface for Agda.Primitive.
+  loadPrims <- optLoadPrimitives <$> pragmaOptions
+  if loadPrims
+    then do
+    [agdaPrimInter] <- filter (("Agda.Primitive"==) . prettyShow . iModuleName)
+      . map miInterface . Map.elems
+        <$> getVisitedModules
+    flip evalStateT Set.empty $ mappend <$> doCompile' f NotMain agdaPrimInter <*> doCompile' f isMain i
```

Comment:
Thanks for the fix!
Here is some code duplication that can be avoided by pulling out the `mappend <$> doCompile' f NotMain agdaPrimInter` into an auxiliary function.  
I'll push a commit here so that you can prepare your talk in peace...